### PR TITLE
Initial port of `LinalgTensorCodegenDriver`s into IREE.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
This PR pulls in the codegen drivers from IREE LLVM sandbox
(https://github.com/google/iree-llvm-sandbox). This does not hook into
the IREE backend yet, but just pulls it, builds and registers it. Adds
a few tests to make sure this works.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
This requires another round of integration to get the latest LLVM commit.

File: iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+      getTileAndFuseOptionsFromConfig(funcOp, tilingLevel);
+  if (failed(defaultTilingOptions)) {
+    return signalPassFailure();
+  }
+  LinalgTilingAndFusionOptions tilingOptions = defaultTilingOptions.getValue();
+  bool doTiling = !tilingOptions.tileSizes.empty();
+  if (!tileSizes.empty()) {
+    doTiling = true;
+    tilingOptions.tileSizes = {tileSizes.begin(), tileSizes.end()};
+  }
```
Comment (hanhanW) [inline iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:132]: Maybe rewrite it as this way which makes it cleaner:

```cpp
bool doTiling = !tileSizes.empty();
if (doTiling) {
  tilingOptions.tileSizes = ...
}
```

File: iree/compiler/Codegen/Sandbox/Passes.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+//===----------------------------------------------------------------------===//
+// IREE specific pass creation methods to allow invocation from within IREEs
+// backend pipelines
+//===----------------------------------------------------------------------===//
+
+/// Creates a pass to drive tile + fuse transformations.
+std::unique_ptr<OperationPass<FuncOp>> createLinalgFusePass(
+    int64_t tilingLevel);
+
+/// Creates a pass to drive transformations on Linalg on tensors. Passing the
```
Comment (hanhanW) [inline iree/compiler/Codegen/Sandbox/Passes.h:32]: The comment is not finished? `Passing the ...`

File: iree/compiler/Codegen/Sandbox/Passes.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+namespace mlir {
+
+/// Creates a pass to drive tile + fuse transformations.
+std::unique_ptr<OperationPass<FuncOp>> createLinalgFusePass();
+
+/// Creates a pass to drive transformations on Linalg on tensors.
+std::unique_ptr<OperationPass<FuncOp>> createLinalgSingleTilingExpertPass();
+
+/// Creates a pass to driver the lowering of vector operations.
+std::unique_ptr<OperationPass<FuncOp>> createLinalgVectorLoweringPass();
```
Comment (hanhanW) [inline iree/compiler/Codegen/Sandbox/Passes.h:21]: Maybe add a block comments for these passes? Are they mirrored from sandbox passes?

File: iree/compiler/Codegen/Sandbox/Passes.td [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      "Set the maximum vector load/store rank.">
+  ];
+  let dependentDialects = [
+    "::mlir::arith::ArithmeticDialect", "::mlir::AffineDialect",
+    "::mlir::memref::MemRefDialect", "::mlir::scf::SCFDialect",
+    "::mlir::StandardOpsDialect", "::mlir::vector::VectorDialect"
+  ];
+}
+
+#endif // IREE_CODEGEN_SANDBOX_PASSES_TD
```
Comment (hanhanW) [inline iree/compiler/Codegen/Sandbox/Passes.td:171]: add a newline at end of file

image

File: iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+std::unique_ptr<OperationPass<FuncOp>> mlir::createLinalgVectorLoweringPass() {
+  return std::make_unique<LinalgVectorLoweringPass>();
+}
+
+//===----------------------------------------------------------------------===//
+// IREE specific pass creation methods to allow invocation from within IREEs
+// backend pipelines
+//===----------------------------------------------------------------------===//
+
+/// Creates a pass to drive transformations on Linalg on tensors.
```
Comment (hanhanW) [inline iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:325]: Delete the comment. It's documented in `.h` file.

File: iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+std::unique_ptr<OperationPass<FuncOp>> createLinalgFusePass(
+    int64_t tilingLevel) {
+  return std::make_unique<LinalgSingleTilingExpertPass>(tilingLevel);
+}
+
+/// Creates a pass to drive transformations on Linalg on tensors.
```
Comment (hanhanW) [inline iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:331]: ditto

**Review by hanhanW — APPROVED** [review]

File: iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:137]: I am not sure that is the same logic. Tiling is needed if there were tile sizes recovered by looking at the `lowering.config` attributes. Then the `tileSizes` from command line, override that. Seems like what you have has a slightly different logic (I dont fully follow it actually).

File: iree/compiler/Codegen/Sandbox/Passes.h (hunk 2 above)
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Sandbox/Passes.h:41]: Sorry, cleaned up the comments.

File: iree/compiler/Codegen/Sandbox/Passes.h (hunk 3 above)
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Sandbox/Passes.h]: I cleaned up the comments. If they arent what you are looking for can you give me an idea of what exactly you are looking for the comments to convey?

**Pull request merged**